tfprotov5/tftypes/attribute_path.go

```diff
@@ -26,6 +28,79 @@ type AttributePath struct {
 	Steps []AttributePathStep
 }
 
+func (a AttributePath) String() string {
+	var res strings.Builder
+	for pos, step := range a.Steps {
+		if pos != 0 {
+			res.WriteString(".")
+		}
+		switch v := step.(type) {
+		case AttributeName:
+			res.WriteString(`AttributeName("` + string(v) + `")`)
+		case ElementKeyString:
+			res.WriteString(`ElementKeyString("` + string(v) + `")`)
+		case ElementKeyInt:
+			res.WriteString(`ElementKeyInt(` + strconv.FormatInt(int64(v), 10) + `)`)
+		case ElementKeyValue:
+			res.WriteString(`ElementKeyValue(` + Value(v).String() + `)`)
+		}
+	}
+	return res.String()
+}
+
+// Equal returns true if two AttributePaths should be considered equal.
+// AttributePaths are considered equal if they have the same number of steps,
+// the steps are all the same types, and the steps have all the same values.
+func (a AttributePath) Equal(o AttributePath) bool {
+	if len(a.Steps) != len(o.Steps) {
+		return false
+	}
+	for pos, aStep := range a.Steps {
+		oStep := o.Steps[pos]
+		switch aVal := aStep.(type) {
+		case AttributeName:
+			oVal, ok := oStep.(AttributeName)
+			if !ok {
+				return false
+			}
+			if oVal != aVal {
+				return false
+			}
+		case ElementKeyString:
+			oVal, ok := oStep.(ElementKeyString)
+			if !ok {
+				return false
+			}
+			if oVal != aVal {
+				return false
+			}
+		case ElementKeyInt:
+			oVal, ok := oStep.(ElementKeyInt)
+			if !ok {
+				return false
+			}
+			if oVal != aVal {
+				return false
+			}
```

Comment:
I think you could probably just collapse these down in to something like the below? The casting seems unnecessary and already handled by the equals? If value's aren't comparable though makes sense to pull that case out.

```go
default:
  if aStep != oStep {
    return false
  }
```